heron/metricsmgr/src/java/com/twitter/heron/metricsmgr/sink/AbstractWebSink.java

```diff
@@ -0,0 +1,165 @@
+
+import com.twitter.heron.common.basics.TypeUtils;
+import com.twitter.heron.spi.metricsmgr.sink.IMetricsSink;
+import com.twitter.heron.spi.metricsmgr.sink.SinkContext;
+
+
+/**
+ * A metrics sink that publishes metrics on a http endpoint
+ */
+abstract class AbstractWebSink implements IMetricsSink {
+  private static final Logger LOG = Logger.getLogger(AbstractWebSink.class.getName());
+
+  private static final int HTTP_STATUS_OK = 200;
+
+  // Metrics will be published on http://host:port/path, the port
+  private static final String KEY_PORT = "port";
+
+  // If you want to specify a file from which to read the port instead of
+  // supplying it directly
+  private static final String KEY_PORT_FILE = "port-file";
+
+  // The path
+  private static final String KEY_PATH = "path";
+
+  // Maximum number of metrics getting served
+  private static final String KEY_METRICS_CACHE_MAX_SIZE = "metrics-cache-max-size";
+  private static final long DEFAULT_MAX_CACHE_SIZE = 1000000;
+
+  // Time To Live before a metric gets evicted from the cache
+  private static final String KEY_METRICS_CACHE_TTL_SEC = "metrics-cache-ttl-sec";
+  private static final long DEFAULT_CACHE_TTL_SECONDS = 600;
+
+  private HttpServer httpServer;
+  private String topologyName;
+  private long cacheMaxSize;
+  private long cacheTtlSeconds;
+  private final Ticker cacheTicker;
+
+  AbstractWebSink() {
+    this(Ticker.systemTicker());
+  }
+
+  @VisibleForTesting
+  AbstractWebSink(Ticker cacheTicker) {
+    this.cacheTicker = cacheTicker;
+  }
+
+  @Override
+  public final void init(Map<String, Object> conf, SinkContext context) {
+    String path = (String) conf.get(KEY_PATH);
+    String portFile = (String) conf.get(KEY_PORT_FILE);
+
+    cacheMaxSize = TypeUtils.getLong(conf.getOrDefault(KEY_METRICS_CACHE_MAX_SIZE,
+        DEFAULT_MAX_CACHE_SIZE));
+
+    cacheTtlSeconds = TypeUtils.getLong(conf.getOrDefault(KEY_METRICS_CACHE_TTL_SEC,
+        DEFAULT_CACHE_TTL_SECONDS));
+
+    topologyName = context.getTopologyName();
+
+    // initialize child classes
+    initialize(conf, context);
+
+    int port = TypeUtils.getInteger(conf.getOrDefault(KEY_PORT, 0));
+    if (port == 0) {
+      if (!Strings.isNullOrEmpty(portFile)) {
+        try {
+          port = TypeUtils.getInteger(Files.lines(Paths.get(portFile)).findFirst()
+              .get().trim());
+        } catch (IOException | SecurityException | IllegalArgumentException e) {
+          throw new IllegalArgumentException("Could not parse " + KEY_PORT_FILE + " " + portFile
+              + " Make sure the file is readable,"
+              + " only contains the port on which the service should run"
+              + " and is UTF8 encoded", e);
+        }
+      } else {
+        throw new IllegalArgumentException("Neither 'port' nor 'port_file' "
+            + "were specified in config for metrics sink " + context.getSinkId());
+      }
+    }
+    startHttpServer(path, port);
```

Comment:
what is this httpserver for ? who will send request to this server?

Reply:
It is there from the old WebSink sink class. This provides an endpoint for the prometheus service to scrape metrics from.  I added this class so both the WebSink and PrometheusSink could share the http server code. Thoughts?

Reply:
I though MetricsSink is a client which send metrics to Prometheus server.
If we need to make MetricsSink as server, why not use tmaster/metricscache httpserver https://github.com/twitter/heron/blob/master/heron/tmaster/src/cpp/manager/stats-interface.h#L43 instead?

Reply:
The metrics for prometheus need to be in a certain format for exporting (https://prometheus.io/docs/instrumenting/exposition_formats/). This why the http server is used. I think keeping it here makes it less invasive with the core code and less stress on the tmaster service. Thoughts?

Reply:
It might also be a more scalable way of doing things because now you have prometheus server connecting to each metrics manager to get the stats so the load is more evenly distributed.

Reply:
@huijunw do you have any more concerns? we would like to get this in. thanks.

Reply:
@cckellogg no more concerns, LGTM